JSDK-2561" wait for initial DLTS handshake before handling glare

makarandp0 opened the pull request:
We did not handle glare when `this._descriptionRevision === 1` - This ensured that DTLS role was negotiated. But that works only for 1st peer connection, when a new peer connection is created in response to VMS failover - the `descriptionRevision` would not be 1. To ensure that every new peerConnection does not handle glare before DLTS negotiation is finished, we will wait for `this._lastStableDescriptionRevision` to be set.


**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

syerrapragada commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -955,7 +954,20 @@ class PeerConnectionV2 extends StateMachine {
+          // }
+          // with glare handling (if dtls role was not negotiated before ) the generated answer will set setup:active.
+          // we do not want that. lets wait for "initial negotiation" before attempting glare handling.
+          if (this._needsAnswer && this._lastStableDescriptionRevision === 0) {
```
Question: Would `this._lastStableDescriptionRevision` be 0 after VMS Failover if at least 1 SDP negotiation already happened? I do not see it being reset to 0 anywhere in the code. More unit/integration tests would help find the answer, I guess

makarandp0 replied:
Yes. For new PC, we set this 0.  https://github.com/twilio/twilio-video.js/blob/f300dc0e161b09d4e77d85405b2c95e7834f529c/lib/signaling/v2/peerconnection.js#L207, 

it will be updated only after new PC completes a negotiation.

makarandp0 replied:
Good point. @innerverse - is revision 0 in RSP a possibility?

innerverse replied:
We always start `create-offer` at rev1.  You should be fine with this.

ceaglest commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -955,7 +954,20 @@ class PeerConnectionV2 extends StateMachine {
+          // }
+          // with glare handling (if dtls role was not negotiated before ) the generated answer will set setup:active.
+          // we do not want that. lets wait for "initial negotiation" before attempting glare handling.
+          if (this._needsAnswer && this._lastStableDescriptionRevision === 0) {
```
Yes. We had the same bug in C++ where we interpreted revision 1 to mean "DTLS role negotiation" but this had a whole host of problems. We now have a boolean to track the DTLS negotiation process that is independent of revision. I think the usage of _lastStableDescriptionRevision would be equivalent here. 👍

makarandp0 commented:
@ceaglest - I looked at adding more unit tests, but it involves too much code refactoring to be worth it. I think the tests that I have already added are sufficient because we have pretty good test coverage for all new and existing code  (https://3920-46595751-gh.circle-artifacts.com/0/home/circleci/project/coverage/lcov-report/lib/signaling/v2/peerconnection.js.html)
 
But while looking at the tests I realized the pattern we are using for unit tests here (`combinationContext` is causing the test time to grow exponentially. And now causing failures on travis. I have now added some checks to reduce redundant tests.

ceaglest commented on test/unit/spec/signaling/v2/peerconnection.js:
```diff
@@ -874,7 +898,18 @@ describe('PeerConnectionV2', () => {
+          // in case of vms fail-over, new PC get to 'have-local-offer' state
+          // by VMS sending create-offer message. Which ends up setting a
+          // test.pcv2._descriptionRevision. lets simulate that.
+          lastOfferRevision = 25; // even though its 'initial' state - last offer will not be 1.
```
What if `lastOfferRevision = 0`?

makarandp0 replied:
@ceaglest - @innerverse confirmed that RSP starts at revision 0, so we should be safe with assumptions here.
